localstack-core/localstack/services/s3/utils.py

```diff
@@ -403,6 +404,43 @@ def parse_copy_source_range_header(copy_source_range: str, object_size: int) ->
     )
 
 
+def get_failed_upload_part_copy_source_preconditions(
```

Comment:
Nice! This looks similar to `get_failed_precondition_copy_source`, and I wonder if it could be consolidated into one. It seems you went a step further and added more parity than what we had before for `CopyObject`, which is great! 

Do you know if those 2 could be consolidated, or do they have behavior differences? The consolidation would be out of scope of the PR as it would require additional tests for both `CopyObject` and `UploadPartCopy`, but it's just for us to know 😄